flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionReader.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * The reader (read view) of a BoundedBlockingSubpartition.
+ */
+class BoundedBlockingSubpartitionReader implements ResultSubpartitionView {
+
+	/** The result subpartition that we read. */
+	private final BoundedBlockingSubpartition parent;
+
+	/** Further byte buffers, to handle cases where there is more data than fits into
+	 * one mapped byte buffer (2GB = Integer.MAX_VALUE). */
+	private final Iterator<ByteBuffer> furtherData;
+
+	/** The reader/decoder to the memory mapped region with the data we currently read from.
+	 * Max 2GB large. Further regions may be in the {@link #furtherData} field.
+	 * Null once the reader empty or disposed.*/
+	@Nullable
+	private BoundedBlockingSubpartitionMemory.Reader data;
+
+	/** The next buffer (look ahead). Null once the data is depleted or reader is disposed. */
+	@Nullable
+	private Buffer nextBuffer;
+
+	/** The remaining number of data buffers (not events) in the result. */
+	private int bufferBacklog;
+
+	/** Flag whether this reader is released. Atomic, to avoid double release. */
+	private boolean isReleased;
+
+	/**
+	 * Convenience constructor that takes a single buffer.
+	 */
+	BoundedBlockingSubpartitionReader(
+			BoundedBlockingSubpartition parent,
+			ByteBuffer data,
+			int numBuffers) throws IOException {
+		this(parent, Collections.singleton(data), numBuffers);
+	}
+
+	/**
+	 * The general constructor that takes a queue of buffers containing the data.
+	 */
+	BoundedBlockingSubpartitionReader(
+			BoundedBlockingSubpartition parent,
+			Iterable<ByteBuffer> data,
+			int numBuffers) throws IOException {
+
+		checkArgument(numBuffers >= 0);
+
+		this.parent = checkNotNull(parent);
+		this.bufferBacklog = numBuffers;
+
+		this.furtherData = data.iterator();
+		checkArgument(furtherData.hasNext());
+
+		this.data = new BoundedBlockingSubpartitionMemory.Reader(furtherData.next());
+		this.nextBuffer = fillNextBuffer();
+	}
+
+	@Nullable
+	@Override
+	public BufferAndBacklog getNextBuffer() throws IOException {
+		final Buffer current = nextBuffer; // copy reference to stack
+
+		if (current == null) {
+			// as per contract, we must return null when the reader is empty,
+			// but also in case the reader is disposed (rather than throwing an exception)
+			return null;
+		}
+		if (current.isBuffer()) {
+			bufferBacklog--;
+		}
+
+		Buffer next = fillNextBuffer();
+		nextBuffer = next;
+
+		return new BufferAndBacklog(
+				current,
+				next != null,
+				bufferBacklog,
+				next != null && !next.isBuffer());
+	}
+
+	@Nullable
+	private Buffer fillNextBuffer() throws IOException {
+		// should only be null once empty or disposed, in which case this method
+		// should not be called any more
+		assert data != null;
+
+		final Buffer next = data.sliceNextBuffer();
+		if (next != null) {
+			return next;
+		}
+
+		if (!furtherData.hasNext()) {
+			data = null;
+			return null;
+		}
+
+		data = new BoundedBlockingSubpartitionMemory.Reader(furtherData.next());
+		return fillNextBuffer();
```

Comment:
Is it recursively to find next buffer untill it is not null? We could use a simple for-loop to avoid temporary `BoundedBlockingSubpartitionMemory.Reader` creation.

Reply:
This is at buffer rollover, which once per 2GB. Code simplicity is more important in that context over avoiding a single object creation.